File: project-pipeline/docs/api-examples/seed_elaboration.py
#!/usr/bin/env python3
"""
=============================================================================
Seed Elaboration API - Python Examples
=============================================================================
This module demonstrates how to interact with the Seed Elaboration API
using Python with the requests library.

Prerequisites:
- Python 3.8+
- requests library: pip install requests

Usage:
    from seed_elaboration import SeedElaborationClient

    client = SeedElaborationClient('your-jwt-token')
    client.run_full_elaboration_flow('seed-uuid')

CLI Usage:
    export JWT_TOKEN='your-token-here'
    export SEED_ID='seed-uuid-here'
    python seed_elaboration.py
=============================================================================
"""

import os
import sys
from typing import Dict, List, Optional, Any
import requests
import json


class SeedElaborationClient:
    """
    Client for interacting with the Seed Elaboration API.

    Attributes:
        jwt_token (str): JWT authentication token
        base_url (str): API base URL
    """

    def __init__(self, jwt_token: str, base_url: str = 'http://localhost:3000'):
        """
        Initialize the Seed Elaboration API client.

        Args:
            jwt_token: JWT authentication token
            base_url: API base URL (default: http://localhost:3000)
        """
        self.jwt_token = jwt_token
        self.base_url = base_url.rstrip('/')
        self.headers = {
            'Content-Type': 'application/json',
            'Authorization': f'Bearer {jwt_token}',
        }

    def start_session(self, seed_id: str) -> Dict[str, Any]:
        """
        Start a new elaboration session.

        Args:
            seed_id: Seed UUID

        Returns:
            StartSessionResponse containing sessionId, question, completeness, etc.

        Raises:
            requests.HTTPError: If the API request fails
        """
        url = f'{self.base_url}/seeds/{seed_id}/elaborate/start'
        response = requests.post(url, headers=self.headers)
        response.raise_for_status()
        return response.json()

    def submit_answer(self, seed_id: str, session_id: str, answer: str) -> Dict[str, Any]:
        """
        Submit an answer to the current question.

        Args:
            seed_id: Seed UUID
            session_id: Session UUID
            answer: User's answer to the current question

        Returns:
            ProcessAnswerResponse containing nextQuestion, completeness, metadata, etc.

        Raises:
            requests.HTTPError: If the API request fails
        """
        url = f'{self.base_url}/seeds/{seed_id}/elaborate/answer'
        payload = {
            'sessionId': session_id,
            'answer': answer,
        }
        response = requests.post(url, headers=self.headers, json=payload)
        response.raise_for_status()
        return response.json()

    def get_status(self, seed_id: str) -> Dict[str, Any]:
        """
        Get elaboration status and progress.

        Args:
            seed_id: Seed UUID

        Returns:
            ElaborationStatusResponse containing completeness, metadata, missingFields

        Raises:
            requests.HTTPError: If the API request fails
        """
        url = f'{self.base_url}/seeds/{seed_id}/elaborate/status'
        response = requests.get(url, headers=self.headers)
        response.raise_for_status()
        return response.json()

    def run_full_elaboration_flow(self, seed_id: str) -> None:
        """
        Run a complete elaboration flow with example answers.

        Args:
            seed_id: Seed UUID
        """
        print('🚀 Starting seed elaboration flow...\n')

        # Start session
        print('📝 Step 1: Starting session...')
        start_response = self.start_session(seed_id)
        session_id = start_response['sessionId']
        print(f'✅ Session started: {session_id}')
        print(f'❓ Question 1: {start_response["question"]}')
        print(f'📊 Completeness: {start_response["completeness"]}%\n')

        # Example answers
        example_answers = [
            '30 young people aged 18-25 from Turkey, Spain, and Germany',
            'Barcelona, Spain, at the Youth Hostel Barcelona',
            '7 days, from July 15 to July 21, 2024',
            '€15,000 total budget, approximately €500 per participant',
            'Cultural workshops, team building activities, and local excursions',
            'Focus on inclusion and environmental sustainability',
            'Travel insurance required, no visas needed for EU participants',
        ]

        # Submit answers
        question_number = 2
        for answer in example_answers:
            print(f'📝 Step {question_number}: Submitting answer...')
            print(f'💬 Answer: "{answer}"')

            answer_response = self.submit_answer(seed_id, session_id, answer)

            print(f'✅ Answer processed')
            print(f'📊 Completeness: {answer_response["completeness"]}%')

            if 'extractedMetadata' in answer_response and answer_response['extractedMetadata']:
                print(f'📦 Extracted metadata:')
                print(json.dumps(answer_response['extractedMetadata'], indent=2))

            if answer_response.get('complete', False):
                print('🎉 Elaboration complete!\n')
                break

            if answer_response.get('nextQuestion'):
                print(f'❓ Next question: {answer_response["nextQuestion"]}')
                if answer_response.get('suggestions'):
                    print(f'💡 Suggestions: {", ".join(answer_response["suggestions"])}')

            print()
            question_number += 1

        # Get final status
        print('📊 Step Final: Getting final status...')
        status = self.get_status(seed_id)
        print(f'✅ Final completeness: {status["completeness"]}%')
        print(f'📦 Final metadata:')
        print(json.dumps(status['metadata'], indent=2))

        if status['missingFields']:
            print(f'⚠️  Missing fields: {", ".join(status["missingFields"])}')
        else:
            print('✅ All required information collected!')


# =============================================================================
# Usage Examples
# =============================================================================

def example1_basic_usage():
    """Example 1: Basic usage - Start session and submit one answer"""
    client = SeedElaborationClient('your-jwt-token-here')
    seed_id = '550e8400-e29b-41d4-a716-446655440000'

    try:
        # Start session
        start_response = client.start_session(seed_id)
        print('First question:', start_response['question'])

        # Submit answer
        answer_response = client.submit_answer(
            seed_id,
            start_response['sessionId'],
            '30 young people aged 18-25 from Turkey, Spain, and Germany'
        )

        print('Next question:', answer_response.get('nextQuestion'))
        print('Completeness:', str(answer_response['completeness']) + '%')
    except requests.HTTPError as e:
        print(f'Error: {e}')
        print(f'Response: {e.response.text}')


def example2_check_status():
    """Example 2: Check status before continuing"""
    client = SeedElaborationClient('your-jwt-token-here')
    seed_id = '550e8400-e29b-41d4-a716-446655440000'

    try:
        # Check current status
        status = client.get_status(seed_id)
        print('Current completeness:', str(status['completeness']) + '%')
        print('Missing fields:', status['missingFields'])

        # Resume if not complete
        if status['completeness'] < 100:
            start_response = client.start_session(seed_id)
            print('Resuming session:', start_response['sessionId'])
            print('Current question:', start_response['question'])
    except requests.HTTPError as e:
        print(f'Error: {e}')
        print(f'Response: {e.response.text}')


def example3_interactive_flow():
    """Example 3: Full elaboration flow with interactive prompts"""
    client = SeedElaborationClient('your-jwt-token-here')
    seed_id = '550e8400-e29b-41d4-a716-446655440000'

    try:
        start_response = client.start_session(seed_id)
        session_id = start_response['sessionId']
        current_question = start_response['question']
        complete = False

        print(f'\n📝 Elaboration started (Session: {session_id})\n')

        while not complete:
            print(f'❓ {current_question}\n')
            answer = input('Your answer: ')

            response = client.submit_answer(seed_id, session_id, answer)

            print(f'\n📊 Completeness: {response["completeness"]}%')

            if response.get('extractedMetadata'):
                print('📦 Extracted:', json.dumps(response['extractedMetadata'], indent=2))

            if response.get('complete', False):
                print('\n🎉 Elaboration complete!\n')
                complete = True
            else:
                current_question = response.get('nextQuestion', '')
                if response.get('suggestions'):
                    print('💡 Suggestions:', ', '.join(response['suggestions']))
                print()
    except requests.HTTPError as e:
        print(f'Error: {e}')
        print(f'Response: {e.response.text}')
    except KeyboardInterrupt:
        print('\n\nElaboration interrupted by user.')


def example4_error_handling():
    """Example 4: Error handling"""
    client = SeedElaborationClient('your-jwt-token-here')
    seed_id = '550e8400-e29b-41d4-a716-446655440000'

    try:
        start_response = client.start_session(seed_id)

        # Try to submit with invalid session ID
        try:
            client.submit_answer(seed_id, 'invalid-session-id', 'Some answer')
        except requests.HTTPError as e:
            print('Expected error for invalid session:', e.response.json()['message'])

        # Submit with correct session ID
        answer_response = client.submit_answer(
            seed_id,
            start_response['sessionId'],
            '30 young people'
        )

        print('Answer submitted successfully')
    except requests.HTTPError as e:
        print(f'Unexpected error: {e}')
        print(f'Response: {e.response.text}')


def example5_batch_processing():
    """Example 5: Process multiple seeds in batch"""
    client = SeedElaborationClient('your-jwt-token-here')
    seed_ids = [
        '550e8400-e29b-41d4-a716-446655440000',
        '550e8400-e29b-41d4-a716-446655440001',
        '550e8400-e29b-41d4-a716-446655440002',
    ]

    for seed_id in seed_ids:
        try:
            print(f'\n📦 Processing seed: {seed_id}')
            status = client.get_status(seed_id)
            print(f'  Completeness: {status["completeness"]}%')

            if status['completeness'] < 100:
                print(f'  Missing fields: {", ".join(status["missingFields"])}')
        except requests.HTTPError as e:
            print(f'  Error: {e.response.json()["message"]}')


# =============================================================================
# CLI Interface
# =============================================================================

def main():
    """Main CLI entry point"""
    jwt_token = os.environ.get('JWT_TOKEN')
    seed_id = os.environ.get('SEED_ID', '550e8400-e29b-41d4-a716-446655440000')

    if not jwt_token:
        print('❌ Error: JWT_TOKEN environment variable is required')
        print('Usage: JWT_TOKEN=your-token SEED_ID=seed-uuid python seed_elaboration.py')
        sys.exit(1)

    client = SeedElaborationClient(jwt_token)

    try:
        client.run_full_elaboration_flow(seed_id)
    except requests.HTTPError as e:
        print(f'\n❌ API Error: {e}')
        try:
            error_detail = e.response.json()
            print(f'Message: {error_detail.get("message", "Unknown error")}')
        except:
            print(f'Response: {e.response.text}')
        sys.exit(1)
    except KeyboardInterrupt:
        print('\n\nProcess interrupted by user.')
        sys.exit(0)


if __name__ == '__main__':
    main()

File: project-pipeline/docs/api-examples/seed-elaboration.js
/**
 * =============================================================================
 * Seed Elaboration API - JavaScript/TypeScript Examples
 * =============================================================================
 * This module demonstrates how to interact with the Seed Elaboration API
 * using JavaScript/TypeScript with the Fetch API.
 *
 * Prerequisites:
 * - Node.js 18+ or browser environment
 * - Valid JWT authentication token
 * - Valid seed UUID
 *
 * Usage (Node.js):
 *   const SeedElaborationClient = require('./seed-elaboration.js');
 *   const client = new SeedElaborationClient('your-jwt-token', 'http://localhost:3000');
 *   await client.runFullElaborationFlow('seed-uuid');
 *
 * Usage (TypeScript):
 *   import { SeedElaborationClient } from './seed-elaboration';
 *   const client = new SeedElaborationClient('your-jwt-token', 'http://localhost:3000');
 *   await client.runFullElaborationFlow('seed-uuid');
 * =============================================================================
 */

// TypeScript interfaces (use these in .ts files)
/**
 * @typedef {Object} SeedMetadata
 * @property {number} [participantCount]
 * @property {string[]} [participantCountries]
 * @property {Object} [ageRange]
 * @property {number} [duration]
 * @property {string} [startDate]
 * @property {string} [endDate]
 * @property {number} [totalBudget]
 * @property {Object} [destination]
 * @property {number} completeness
 */

/**
 * @typedef {Object} StartSessionResponse
 * @property {string} sessionId
 * @property {string} question
 * @property {number} [questionNumber]
 * @property {number} [totalQuestions]
 * @property {number} completeness
 * @property {string[]} [suggestions]
 * @property {SeedMetadata} metadata
 */

/**
 * @typedef {Object} ProcessAnswerResponse
 * @property {string} sessionId
 * @property {string|null} nextQuestion
 * @property {number} [questionNumber]
 * @property {number} [totalQuestions]
 * @property {number} completeness
 * @property {boolean} complete
 * @property {Object} [extractedMetadata]
 * @property {string[]} [suggestions]
 * @property {string[]} [validationErrors]
 * @property {SeedMetadata} metadata
 */

/**
 * @typedef {Object} ElaborationStatusResponse
 * @property {number} completeness
 * @property {SeedMetadata} metadata
 * @property {string[]} missingFields
 */

class SeedElaborationClient {
  /**
   * Create a new Seed Elaboration API client
   * @param {string} jwtToken - JWT authentication token
   * @param {string} [baseUrl='http://localhost:3000'] - API base URL
   */
  constructor(jwtToken, baseUrl = 'http://localhost:3000') {
    this.jwtToken = jwtToken;
    this.baseUrl = baseUrl;
    this.headers = {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${jwtToken}`,
    };
  }

  /**
   * Start a new elaboration session
   * @param {string} seedId - Seed UUID
   * @returns {Promise<StartSessionResponse>}
   */
  async startSession(seedId) {
    const response = await fetch(`${this.baseUrl}/seeds/${seedId}/elaborate/start`, {
      method: 'POST',
      headers: this.headers,
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(`Failed to start session: ${error.message}`);
    }

    return await response.json();
  }

  /**
   * Submit an answer to the current question
   * @param {string} seedId - Seed UUID
   * @param {string} sessionId - Session UUID
   * @param {string} answer - User's answer
   * @returns {Promise<ProcessAnswerResponse>}
   */
  async submitAnswer(seedId, sessionId, answer) {
    const response = await fetch(`${this.baseUrl}/seeds/${seedId}/elaborate/answer`, {
      method: 'POST',
      headers: this.headers,
      body: JSON.stringify({ sessionId, answer }),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(`Failed to submit answer: ${error.message}`);
    }

    return await response.json();
  }

  /**
   * Get elaboration status
   * @param {string} seedId - Seed UUID
   * @returns {Promise<ElaborationStatusResponse>}
   */
  async getStatus(seedId) {
    const response = await fetch(`${this.baseUrl}/seeds/${seedId}/elaborate/status`, {
      method: 'GET',
      headers: this.headers,
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(`Failed to get status: ${error.message}`);
    }

    return await response.json();
  }

  /**
   * Run a complete elaboration flow with example answers
   * @param {string} seedId - Seed UUID
   * @returns {Promise<void>}
   */
  async runFullElaborationFlow(seedId) {
    console.log('🚀 Starting seed elaboration flow...\n');

    // Start session
    console.log('📝 Step 1: Starting session...');
    const startResponse = await this.startSession(seedId);
    console.log(`✅ Session started: ${startResponse.sessionId}`);
    console.log(`❓ Question 1: ${startResponse.question}`);
    console.log(`📊 Completeness: ${startResponse.completeness}%\n`);

    const sessionId = startResponse.sessionId;
    const exampleAnswers = [
      '30 young people aged 18-25 from Turkey, Spain, and Germany',
      'Barcelona, Spain, at the Youth Hostel Barcelona',
      '7 days, from July 15 to July 21, 2024',
      '€15,000 total budget, approximately €500 per participant',
      'Cultural workshops, team building activities, and local excursions',
      'Focus on inclusion and environmental sustainability',
      'Travel insurance required, no visas needed for EU participants',
    ];

    // Submit answers
    let questionNumber = 2;
    for (const answer of exampleAnswers) {
      console.log(`📝 Step ${questionNumber}: Submitting answer...`);
      console.log(`💬 Answer: "${answer}"`);

      const answerResponse = await this.submitAnswer(seedId, sessionId, answer);

      console.log(`✅ Answer processed`);
      console.log(`📊 Completeness: ${answerResponse.completeness}%`);

      if (answerResponse.extractedMetadata) {
        console.log(`📦 Extracted metadata:`, JSON.stringify(answerResponse.extractedMetadata, null, 2));
      }

      if (answerResponse.complete) {
        console.log('🎉 Elaboration complete!\n');
        break;
      }

      if (answerResponse.nextQuestion) {
        console.log(`❓ Next question: ${answerResponse.nextQuestion}`);
        if (answerResponse.suggestions && answerResponse.suggestions.length > 0) {
          console.log(`💡 Suggestions: ${answerResponse.suggestions.join(', ')}`);
        }
      }

      console.log('');
      questionNumber++;
    }

    // Get final status
    console.log('📊 Step Final: Getting final status...');
    const status = await this.getStatus(seedId);
    console.log(`✅ Final completeness: ${status.completeness}%`);
    console.log(`📦 Final metadata:`, JSON.stringify(status.metadata, null, 2));

    if (status.missingFields.length > 0) {
      console.log(`⚠️  Missing fields: ${status.missingFields.join(', ')}`);
    } else {
      console.log('✅ All required information collected!');
    }
  }
}

// =============================================================================
// Usage Examples
// =============================================================================

/**
 * Example 1: Basic usage - Start session and submit one answer
 */
async function example1_basicUsage() {
  const client = new SeedElaborationClient('your-jwt-token-here');
  const seedId = '550e8400-e29b-41d4-a716-446655440000';

  try {
    // Start session
    const startResponse = await client.startSession(seedId);
    console.log('First question:', startResponse.question);

    // Submit answer
    const answerResponse = await client.submitAnswer(
      seedId,
      startResponse.sessionId,
      '30 young people aged 18-25 from Turkey, Spain, and Germany'
    );

    console.log('Next question:', answerResponse.nextQuestion);
    console.log('Completeness:', answerResponse.completeness + '%');
  } catch (error) {
    console.error('Error:', error.message);
  }
}

/**
 * Example 2: Check status before continuing
 */
async function example2_checkStatus() {
  const client = new SeedElaborationClient('your-jwt-token-here');
  const seedId = '550e8400-e29b-41d4-a716-446655440000';

  try {
    // Check current status
    const status = await client.getStatus(seedId);
    console.log('Current completeness:', status.completeness + '%');
    console.log('Missing fields:', status.missingFields);

    // Resume if not complete
    if (status.completeness < 100) {
      const startResponse = await client.startSession(seedId);
      console.log('Resuming session:', startResponse.sessionId);
      console.log('Current question:', startResponse.question);
    }
  } catch (error) {
    console.error('Error:', error.message);
  }
}

/**
 * Example 3: Full elaboration flow with interactive prompts (Node.js)
 */
async function example3_interactiveFlow() {
  const readline = require('readline');
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  const client = new SeedElaborationClient('your-jwt-token-here');
  const seedId = '550e8400-e29b-41d4-a716-446655440000';

  try {
    const startResponse = await client.startSession(seedId);
    let sessionId = startResponse.sessionId;
    let currentQuestion = startResponse.question;
    let complete = false;

    console.log(`\n📝 Elaboration started (Session: ${sessionId})\n`);

    while (!complete) {
      console.log(`❓ ${currentQuestion}\n`);

      const answer = await new Promise((resolve) => {
        rl.question('Your answer: ', resolve);
      });

      const response = await client.submitAnswer(seedId, sessionId, answer);

      console.log(`\n📊 Completeness: ${response.completeness}%`);

      if (response.extractedMetadata) {
        console.log('📦 Extracted:', JSON.stringify(response.extractedMetadata, null, 2));
      }

      if (response.complete) {
        console.log('\n🎉 Elaboration complete!\n');
        complete = true;
      } else {
        currentQuestion = response.nextQuestion;
        if (response.suggestions && response.suggestions.length > 0) {
          console.log('💡 Suggestions:', response.suggestions.join(', '));
        }
        console.log('');
      }
    }

    rl.close();
  } catch (error) {
    console.error('Error:', error.message);
    rl.close();
  }
}

/**
 * Example 4: Error handling
 */
async function example4_errorHandling() {
  const client = new SeedElaborationClient('your-jwt-token-here');
  const seedId = '550e8400-e29b-41d4-a716-446655440000';

  try {
    const startResponse = await client.startSession(seedId);

    // Try to submit with invalid session ID
    try {
      await client.submitAnswer(seedId, 'invalid-session-id', 'Some answer');
    } catch (error) {
      console.error('Expected error for invalid session:', error.message);
    }

    // Submit with correct session ID
    const answerResponse = await client.submitAnswer(
      seedId,
      startResponse.sessionId,
      '30 young people'
    );

    console.log('Answer submitted successfully');
  } catch (error) {
    console.error('Unexpected error:', error.message);
  }
}

// Export for use as module
if (typeof module !== 'undefined' && module.exports) {
  module.exports = SeedElaborationClient;
}

// Run example if executed directly
if (require.main === module) {
  const JWT_TOKEN = process.env.JWT_TOKEN;
  const SEED_ID = process.env.SEED_ID || '550e8400-e29b-41d4-a716-446655440000';

  if (!JWT_TOKEN) {
    console.error('❌ Error: JWT_TOKEN environment variable is required');
    console.error('Usage: JWT_TOKEN=your-token SEED_ID=seed-uuid node seed-elaboration.js');
    process.exit(1);
  }

  const client = new SeedElaborationClient(JWT_TOKEN);
  client.runFullElaborationFlow(SEED_ID).catch(console.error);
}
